src/sulogin-shell/sulogin-shell.c

```diff
@@ -103,17 +132,35 @@ int main(int argc, char *argv[]) {
                 /* allows passwordless logins if root account is locked. */
                 sulogin_cmdline[1] = "--force";
 
-        (void) fork_wait(sulogin_cmdline);
+        for (;;) {
+                (void) fork_wait(sulogin_cmdline);
+
+                r = bus_connect_system_systemd(&bus);
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to get D-Bus connection: %m");
+                        continue;
+                }
+
+                r = reload_manager(bus);
+                if (r < 0)
+                        goto close_bus;
 
-        r = bus_connect_system_systemd(&bus);
-        if (r < 0) {
-                log_warning_errno(r, "Failed to get D-Bus connection: %m");
-                r = 0;
-        } else {
-                (void) reload_manager(bus);
+                r = is_inactive_default_target(bus, &is_inactive);
+                if (r < 0)
+                        goto close_bus;
 
-                r = start_default_target(bus);
+                if (is_inactive) {
+                        r = start_default_target(bus);
+                        if (r == 0)
+                                break;
+                }
+
+        close_bus:
+                sd_bus_flush_close_unref(bus);
+                bus = NULL;
```

Comment:
Please declare `bus` with the `_cleanup_` attribute in the loop. Then, this is not necessary, and you can replace the above `goto` with `continue`.

Reply:
Thanks. This is helpful. This part was unsatisfactory also to me.